Class for RadioGroup & LabeledStaticBox and improvements for Calibration Dialogs, Windows with wxStaticBoxSizer

yw4z opened the pull request:
### RADIOGROUP CLASS
![Screenshot-20250604124056](https://github.com/user-attachments/assets/36aed081-1d9f-4e27-bbfa-60d60c07f91d)
• Fully featured RadioButton class for hustle free UI creation. just requires a list
`m_rbFilamentType = new RadioGroup(this, { _L("PLA"), _L("ABS/ASA"), _L("PETG"), _L("PCTG"), _L("TPU"), _L("PA-CF"), _L("PET-CF"), _L("Custom") }, wxVERTICAL, 2);`

• Supports dark mode, keyboard input, hover effects and focus rectangle

• Supports direction for vertical or horizontal orientation

• Supports optional item count limit for row / column. switches to new row / column after given limit reaches depends on choosen direction / orientation

• Supports choosing item with clicking on text

• Matches look between platforms

### LABELEDSTATICBOX CLASS
• Adds new LabeledStaticBox class for modernizing wxStaticBoxSizer with to fix dark mode related problems

### CALIBRATION DIALOGS

**COMMON CHANGES**
• Simplifies and matches code format on all calibration dialogs
• Fixes margin and spacing related problems
• Fixes all dark mode problems
• Matches font size
• Parameter boxes > Matches width
• Parameter boxes > Aligns text to left instead center

**TEMPERATURE CALIBRATION DIALOG**

•  No specific changes

Before-After
![Screenshot-20250604105152](https://github.com/user-attachments/assets/033a12d5-4b57-43b0-9cd4-1ec7fdd70132)


**PRESSURE ADVANCE CALIBRATION DIALOG**

• Used horizontal orientation for radio groups
• aligned checkbox with parameter boxes like on sidebar. ant that required so much changes. thats why i removed create_item_checkbox()

Before-After
![Screenshot-20250604105219](https://github.com/user-attachments/assets/c9ba6603-002c-4aae-8c70-471850ce49f2)


**RETRACTION CALIBRATION DIALOG**

•  Fixed side text on parameter box for "step: ". previously that used "mm/mm". that was confusing
• Used uppercase for first letter on "Step:" .  Not requires new translation

Before-After
![Screenshot-20250604121146](https://github.com/user-attachments/assets/a29d502e-a690-4f78-9229-828c57bfbe5b)

**INPUT SHAPER FREQUENCY TEST DIALOG**

• used single text for X and Y "Start / End" and used parameter boxes alongside like on Printer or Filament settings window. this change requires new translation
• Used lowercase for side text."hz" instead "HZ" . this matches with other parameter boxes

Before-After
![Screenshot-20250604105503](https://github.com/user-attachments/assets/d8033b8d-0a76-4b47-b204-7dca0c3fc0ca)


**INPUT SHAPER DAMP TEST DIALOG**

• used single text for "Frequency X / Y" and used parameter boxes alongside like on Printer or Filament settings window. this change requires new translation
• used single text for "Damp Start / End" and used parameter boxes alongside like on Printer or Filament settings window. this change requires new translation
• Used lowercase for side text."hz" instead "HZ" . this matches with other parameter boxes

Before-After
![Screenshot-20250604105526](https://github.com/user-attachments/assets/695494d3-6765-44f5-8019-a31b3af9aab2)

**MAX VOLUMETRIC SPEED DIALOG**

• Used uppercase for first letter on "Step:" .  Not requires new translation

Before-After
![Screenshot-20250604121248](https://github.com/user-attachments/assets/d522f308-352b-4054-acca-f13738781aaa)

*VFA TEST DIALOG**

• Used uppercase for first letter on "Step:" .  Not requires new translation

Before-After
![Screenshot-20250604121335](https://github.com/user-attachments/assets/441910b5-204a-42e5-8640-15b3afbc5731)

**JUNCTION DEVIATION TEST DIALOG**

•  No specific changes

Before-After
![Screenshot-20250604105636](https://github.com/user-attachments/assets/b288ebb1-d979-48bb-a601-4772d2d22519)


### BED SHAPE DIALOG
Before-After
![Screenshot-20250604100856](https://github.com/user-attachments/assets/ed3441b6-9a5f-4b10-97fe-c341483db9f3)
![Screenshot-20250604100842](https://github.com/user-attachments/assets/1de3bdcc-c69e-4316-a678-8ee4461fb528)

### PHYSICAL PRINTER DIALOG
Before-After
![Screenshot-20250605195001](https://github.com/user-attachments/assets/d26bf8ff-9b77-480c-aab7-09414bd68193)
![Screenshot-20250605194906](https://github.com/user-attachments/assets/9c9f407d-173f-4d18-a0e9-bac6fee6d772)

### UNSAVED CHANGES DIALOG > FullCompareDialog
Couldnt find how to open this dialog

### SAVE PRESET DIALOG
Couldnt find how to open this dialog

ianalexis commented:
at last someone who really understands about gui was able to solve the calibration error.
I had tried to solve it when I developed the InputShaping and Junction Deviation test but without success.
My eyes thank you.

yw4z commented:
thanks still i'm trying to fix main UI elements and is assume 1 passed. icons, margins, colors, components etc.. but UI improved nicely and worth the effort

yw4z commented:
also looks like LabeledStaticBox not rendering correctly. i will check when i got time

Noisyfox commented:
![image](https://github.com/user-attachments/assets/1d6e48a5-910f-4950-896b-0eae0d83748d)

The static box doesn't work well on macOS.

yw4z commented:
should i switch to wxWindow for LabeledStaticBox instead dealing with wxStaticBox. i think border will always be a problem on linux. looks like paint event didnt even used

i assume creating a sizer in that wxWindow and using that will get job done
example from OptionsGroup.cpp
```
LabeledStaticBox * stb = new LabeledStaticBox(m_parent, _(title));
this->stb = stb;
sizer = new wxStaticBoxSizer(stb, wxVERTICAL); 
```
will look like but might require a bit more change on various areas
```
LabeledStaticBox * stb = new LabeledStaticBox(m_parent, _(title));
this->stb = stb;
sizer = new wxBoxSizer(stb->GetSizer() , wxVERTICAL); 
```

i assume dc.clear() also cleans other elements except icons on RadioGroup for linux

yw4z commented:
clears all draw (focus rectangle in this case) and re draws it. but i will try move and resize method instead clearing

Noisyfox commented:
> clears all draw (focus rectangle in this case) and re draws it

Is there any function like `refresh` that do similar thing?

Noisyfox commented:
![linux-radio-focus-highlight](https://github.com/user-attachments/assets/b1a0e26b-0f4e-47b0-940b-11a4b06072b0)

The focus highlight does not work properly on Linux.

yw4z commented:
it catches wxEVT_SET_FOCUS and wxEVT_KILL_FOCUS but looks like m_focus_rect not updating, at least we got better results compared to previous one

yw4z commented:
switched to wxWindow for RadioGroup
i found HasFocus() not working on wxPanel with Linux
also added 2 overrides these ensures focus-ability on linux. let me know if they are unnecessary 
```
    bool AcceptsFocus()             const override { return true; }
    bool AcceptsFocusFromKeyboard() const override { return true; }
```

Noisyfox commented:
The new impl has some minor glitches on Windows:
1. Hovering highlight doesn't work reliably
2. When click the non-selected radio button, the highlight box will shown on the previous selected button for a fraction of second before switching to the clicked button

yw4z commented on src/slic3r/GUI/Widgets/LabeledStaticBox.cpp:
```diff
@@ -0,0 +1,145 @@
+
+bool LabeledStaticBox::Enable(bool enable)
+{
+    bool result = this->Enable(enable);
```
`bool result = this->wxStaticBox::Enable(enable);`

should i use `this->` in here

Noisyfox commented:
macOS:
![image](https://github.com/user-attachments/assets/8f3ebb5b-d82e-4722-bd51-71b98863f577)

yw4z commented:
what a rabbit hole :)

LabeledStaticBox
macOS starts render at wrong position and gets wrong size for rectangle or it draws border outside of rectangle. both easily fixable with some overrides

RadioGroup
i lost about linux. looks like it captures events but not in correct way. probably i cant fix without testing on linux

what you suggest? i dont want to annoy you for each commit

Noisyfox commented:
I'll try doing some debugging on Linux in the next few days.

yw4z commented:
Great, events needs some fine tuning on linux for RadioGroup but it completely ignores painting on LabeledStaticBox. i saw some recommends using GTK styling. not sure how to do that

Noisyfox commented:
> i found this for macOS to edit control margins

You know what:
![image](https://github.com/user-attachments/assets/fe88f66e-6095-4ccb-a610-9620a5f18b72)

wxwidgets actually does the same thing for its staticbox impl. I think we could do that too. I'll give this a quick test.

Noisyfox commented:
Hmmmm wait... You didn't use wxpanel as the parent of the new static box. This should already been applied then?

yw4z commented:
nope i didnt switched that because it broke combability with OptionGroup.cpp. i assume we have to use wxStaticBox to keep things working otherwise we have to fix new problems. 

i read ContentViewMargin defaults to (5,5) after creation maybe we have to override it second time

here i added required changes to a separate branch because changes makes impossible to revert if we want to

https://github.com/yw4z/OrcaSlicer/tree/staticbox-as-window

https://github.com/SoftFever/OrcaSlicer/compare/main...yw4z:OrcaSlicer:staticbox-as-window

Currently i only corrected parents for Junction Deviation Dialog. other dialogs will not work properly

![Screenshot-20250610162610](https://github.com/user-attachments/assets/db44a53c-e3bf-44ef-a9ba-e8087df46542)

i think we have to change parents of controls automatically when they added to sizer to make OptionGroup work
only way i found track sizer changes wxEVT_SIZE event but its not a proper method
other option using a timer to track child controls of sizer but i didnt like it either

Noisyfox commented:
I managed to solve the issue by setting the native NSBox's border width to 0 on macOS:
```
void staticbox_remove_margin(wxStaticBox* sb) {
    NSBox* nativeBox = (NSBox*)sb->GetHandle();
    [nativeBox setBoxType:NSBoxCustom];
    [nativeBox setBorderWidth:0];
}
```
![image](https://github.com/user-attachments/assets/66cdcaec-c666-4c60-93ac-4de1c73c77cc)

And I also have to set `m_pos = {0,0};` too otherwise the left side misses 1px:
![image](https://github.com/user-attachments/assets/7d457392-7175-4122-8c31-9f2282b7e980)

The reason is the position is actually (-1, -1) for default position.

yw4z commented:
looks like macOS adding extra spacing inside staticbox to all edges . normally title and contents aligns on left. looks great anyway

Problems starts to appear when using LabeledStaticBox(based on wxPanel) on OptionGroup

do you think we really need to make that switch because it completely broke controls generated by option group

macOS looks ok but i cant comment about linux

Noisyfox commented:
I think current impl looks good enough, since issues on macOS has now been fixed.

yw4z commented:
yep thats why i asked. should we exclude paint event from linux

yw4z commented:
i can simplify code calibration dialogs with using wxFlexGridSizer for layout. that will remove requirement of calculating text width , simplifies margins and removes unnecessary sizers
```
    auto start_PA_sizer = new wxBoxSizer(wxHORIZONTAL);
    auto start_pa_text = new wxStaticText(this, wxID_ANY, start_pa_str, wxDefaultPosition, st_size, wxALIGN_LEFT);
    m_tiStartPA = new TextInput(this, "", "", "", wxDefaultPosition, ti_size, wxTE_PROCESS_ENTER);
    m_tiStartPA->GetTextCtrl()->SetValidator(wxTextValidator(wxFILTER_NUMERIC));
    start_PA_sizer->Add(start_pa_text, 0, wxALL | wxALIGN_CENTER_VERTICAL, FromDIP(2));
    start_PA_sizer->Add(m_tiStartPA  , 0, wxALL | wxALIGN_CENTER_VERTICAL, FromDIP(2));
    settings_sizer->Add(start_PA_sizer, 0, wxLEFT, FromDIP(3));
```
with wxFlexGridSizer(row_count, 2, FromDIP(4) , FromDIP(20));
```
    auto start_pa_text = new wxStaticText(this, wxID_ANY, start_pa_str);
    m_tiStartPA = new TextInput(this, "", "", "", wxDefaultPosition, ti_size, wxTE_PROCESS_ENTER);
    m_tiStartPA->GetTextCtrl()->SetValidator(wxTextValidator(wxFILTER_NUMERIC));
    f_sizer->Add(start_pa_text, 0, wxEXPAND | wxALIGN_CENTER_VERTICAL);
    f_sizer->Add(m_tiStartPA);
```

Noisyfox commented:
> yep thats why i asked. should we exclude paint event from linux

I haven't looked at that yet. Will give it a look tonight.

yw4z commented:
i read some gtk controls not create paint events. i think it draws natively
is that possible to create a wxWindow in that LabeledStaticBox class and cover staticbox's border and label. i think it has to be click thru and center of it should be masked. ofc its not a good solution

Noisyfox commented:
> is that possible to create a wxWindow in that LabeledStaticBox class and cover staticbox's border and label.

I don't think so because the native box will clip its child.

Noisyfox reviewed (approved):
LGTM, thanks

Noisyfox commented:
Just noticed this on macOS:

![image](https://github.com/user-attachments/assets/0a7ef4ac-cdcd-4982-9405-26ec44b1c01b)

Noisyfox commented:
I think it has something to do with codes like
https://github.com/SoftFever/OrcaSlicer/blob/0999cb057d62d37e783a34f8baa59ac55c150172/src/slic3r/GUI/OptionsGroup.cpp#L264

Some of those `wxOSX ? 0 : xxx` might not needed anymore.

yw4z commented:
please add comments if you figure out which line controls which parts. i have struggled to understand that part. so many conditions used
